File: Labs/Lab12/Lab12_md2.m
% molecular dynamics in two dimensions
%
% Matlab version with keyboard interactions added by H. Nakanishi
%    (based on original TB version by N. Giordano)
% modified by Denes Molnar Nov 2019
%
clear;
clf;
%
% Read parameters
%
Npart  = input('# of particles: ');
L      = input('side length L [units of sigma]: ');
dt     = input('time step dt [units given on p.274]: ');
n_plot = input('plot/record after this many steps: ');
vmax   = input('max initial speed: ');
dmax   = input('max initial relative displacement: ');
out    = input('file name for thermodynamics output: ');
disp('Type c or d (clear/remove green tracks), p to pause, or q to plot time series and end.\n'); 
disp('type + to increase speed 50% or - to decrease 50%, or \n');
disp('type N [N=1,...,4] to increase speed by 10N%. \n');
disp('Hit any key to start...');

x  = zeros(2,Npart);
y  = zeros(2,Npart);
vx = zeros(Npart,1);
vy = zeros(Npart,1);
xp = zeros(Npart,1);
yp = zeros(Npart,1);

%s = RandStream('mt19937ar','Seed',0);
%RandStream.setGlobalStream(s);

%
% Generate initial conditions
%
sqN = sqrt(Npart);
if (sqN == floor(sqN))
   grid = L / sqN;
else
   grid = L / floor(sqN + 1);
end

n = 0;
i = 0;
while i < L       % arrange particles on a roughly square lattice
   j = 0;          % to keep them apart initially
   while j < L
      n = n + 1;
      if n <= Npart
         x(2,n) = i + 0.5 * grid + dmax * (rand-0.5) * grid * sqrt(2);
         y(2,n) = j + 0.5 * grid + dmax * (rand-0.5) * grid * sqrt(2);
         % random initial velocities
         vx(n) = vmax * (rand-0.5) * sqrt(2);
         vy(n) = vmax * (rand-0.5) * sqrt(2);
         x(1,n) = x(2,n) - vx(n) * dt;
         y(1,n) = y(2,n) - vy(n) * dt;
      end
      j = j + grid;
   end
   i = i + grid;
end %i

t = 0;

% initial display of particles
%
symSize = 200; %symbol size (area)
h3 = scatter(x(2,3:Npart), y(2,3:Npart), symSize, 'r', 'Filled');
hold on;
h1 = scatter(x(2,1:1), y(2,1:1), symSize, 'b', 'Filled');
h2 = scatter(x(2,2:2), y(2,2:2), symSize, 'g', 'Filled');

xlabel('x');
ylabel('y');
xlim([0 L]);
ylim([0 L]);
hold on;
pause;

drg = 1;    % whether to draw green circles for positions
for i = 1:Npart
   xp(i) = x(2,i);
   yp(i) = y(2,i);
end
set(gcf, 'CurrentCharacter', '@'); % set to a dummy character



% 
% Evolve with Newtonian dynamics
%
i = 0;
j = 0;
n_p = 0;
r2 = 0;
x1 = 0;
y1 = 0;
x2 = x(2,2) - x(2,1);
y2 = y(2,2) - y(2,1);

while 1    % infinite loop

   [x,y,vx,vy,r2,rp2,x1,y1,x2,y2] = Lab12_updatemd(x, y, vx, vy, Npart, L, dt, x1, y1, x2, y2);

   t = t + dt;
   j = j + 1; 		% use to keep track of how often to plot on screen
   if j >= n_plot		% and record values for later plotting
      if drg == 1
         scatter(xp(1:Npart), yp(1:Npart), symSize, 'g');
      end
      %set(h1, 'XData', x(2,1), 'YData', y(2,1));
      %set(h2, 'XData', x(2,2), 'YData', y(2,2));
      %set(h3, 'XData', x(2,3:Npart), 'YData', y(2,3:Npart));
      %drawnow;
      for k = 1:Npart
         xp(k) = x(2,k);
         yp(k) = y(2,k);
      end

      j = 0;
      i = i + 1;

      % calculate and record time, energy, and temperature
      [E,Epot,T] = Lab12_calc_energy(x, y, vx, vy, Npart, L);
      s = sprintf('t = %0.5f, energy = %0.5e, temperature = %0.5e \n', t, E, T);
      title(s);

      thermo(i,:) = [t, E, T, r2, rp2];
      if out
         dlmwrite(out, thermo(i,:), '-append', 'newline', 'pc', 'delimiter', '\t');
      end
      n_p = n_p + 1;
      velRescale = 1;
      %pause(0.001);
      figure(gcf);
      key = get(gcf, 'CurrentCharacter');
      %key = input('input: ', 's');
      switch key
         case 'q'
            break;
         case {'c','d'}
            clf;
            h3 = scatter(x(2,3:Npart), y(2,3:Npart), symSize, 'r', 'Filled');
            hold on;
            h1 = scatter(x(2,1), y(2,1), symSize, 'b', 'Filled');
            h2 = scatter(x(2,2), y(2,2), symSize, 'g', 'Filled');
            xlabel('x');
            ylabel('y');
            xlim([0 L]);
            ylim([0 L]);
            hold on;
            drawnow;
            if key == 'd'
                drg = 0;
            end
         case 'u'
             drg = 1;
         case 'p'
            pause;
         case '+'   % heat up v -> v * 1.5
            velRescale = 1.5;
         case '-'   % cool v -> v * 0.5
            velRescale = 0.5;
         case '1'  % heat up v -> v * 1.1
            velRescale = 1.1;
         case '2'  % heat up v -> v * 1.2
            velRescale = 1.2;
         case '3'  % heat up v -> v * 1.3
            velRescale = 1.3;
         case '4'  % heat up v -> v * 1.4
            velRescale = 1.4;
      end %switch
      set(gcf, 'CurrentCharacter', '@');
      % rescale velocities, if requested
      if velRescale ~= 1
         for k = 1:Npart
            x(1,k) = x(2,k) - velRescale * (x(2,k) - x(1,k));
            y(1,k) = y(2,k) - velRescale * (y(2,k) - y(1,k));
         end
      end
   end %
   set(gcf, 'CurrentCharacter', '@');
end %while

% now finished      prepare for plotting final results
disp('hit any key to display time series ...')
pause;

lw = 3;

clf;
plot(thermo(:,1), thermo(:,2), 'r', 'LineWidth', lw)
title('Energy time series');
xlabel('Time (in L-J units)');
ylabel('Total energy (in epsilon)');
disp('hit any key to continue')
pause;

clf;
plot(thermo(:,1), thermo(:,3), 'b', 'LineWidth', lw)
title('Temperature time series');
xlabel('Time (in L-J units)');
ylabel('Temperature (in epsilon/k_B)');
disp('hit any key to continue')
pause;

clf;
plot(thermo(:,1), thermo(:,4), 'g', 'LineWidth', lw)
title('Tagged particle square displacement time series');
xlabel('Time (in L-J units)');
ylabel('r^2 (in sigma^2)');
disp('hit any key to continue')
pause;

clf;
plot(thermo(:,1), thermo(:,5), 'g', 'LineWidth', lw)
title('Tagged pair square distance time series');
xlabel('Time (in L-J units)');
ylabel('dr^2 (in sigma^2)');
%
%
%END

File: Labs/Lab12/lab12-md2.py
# molecular dynamics in 2D
#
# - Lennard-Jones potential
# - reduced (dimensionless) quantities
# - LxL box with periodic boundary conditions
# - initconds: random velocity
#              particles on a grid + a random displacement
#
#
# by Denes Molnar for PHYS 580, Fall 2020



import numpy as np
from random import random
from math import sqrt

import matplotlib
matplotlib.use('TkAgg')
import matplotlib.pyplot as plt


# sign function
def sgn(x):
   if x < 0.:    return -1.
   elif x > 0.:  return 1.
   else:         return 0.


# compute one timestep
def update(r, rprev, params):
   # unpack params
   (N, L, dt, rcut) = params

   # storage for updated position and current velocity
   rnew = np.zeros( (N, 2) )
   v = np.zeros( (N, 2) )

   # Verlet update
   for i in range(N):
      # compute total force
      fxTot, fyTot = 0., 0.
      for j in range(N):
         if j == i: continue   # no self-interaction
         # relative position - take nearest separation (periodic BC)
         xij = r[i,0] - r[j,0]
         yij = r[i,1] - r[j,1]
         if abs(xij) > L * 0.5:
           xij -= sgn(xij) * L
         if abs(yij) > L * 0.5:
           yij -= sgn(yij) * L
         rij = sqrt(xij**2 + yij**2)
         # skip r > rcut
         if rij > rcut: continue
         # force j->i
         fjiMagnitude = (48. / rij**13 - 24. / rij**7)
         fxTot += fjiMagnitude * xij / rij
         fyTot += fjiMagnitude * yij / rij
      # new positions and velocities from Verlet
      xnew = 2. * r[i,0] - rprev[i,0] + dt**2 * fxTot
      ynew = 2. * r[i,1] - rprev[i,1] + dt**2 * fyTot
      v[i,0] = (xnew - rprev[i,0]) / (2. * dt)
      v[i,1] = (ynew - rprev[i,1]) / (2. * dt)
      # employ periodic BC, then store position
      if   xnew < 0.:
         xnew += L
         r[i,0] += L
      elif xnew > L:
         xnew -= L
         r[i,0] -= L
      if   ynew < 0.:
         ynew += L
         r[i,1] += L
      elif ynew > L:
         ynew -= L
         r[i,1] -= L
      rnew[i,0] = xnew
      rnew[i,1] = ynew
   # return new position, current position, and current velocity arrays
   return rnew, r, v


# calculate energy, and estimate temperature
def calculateE(r, v, params):

   (N, L, _, rcut) = params

   Ekin, Epot = 0., 0.
   for i in range(N):
      # add kinetic energy
      Ekin += 0.5 * (v[i,0]**2 + v[i,1]**2)
      # for potential energy, take each pair once (e.g., j < i)
      for j in range(i):
         xij = r[i,0] - r[j,0]
         yij = r[i,1] - r[j,1]
         if abs(xij) > L * 0.5:
           xij -= sgn(xij) * L
         if abs(yij) > L * 0.5:
           yij -= sgn(yij) * L
         rij = sqrt(xij**2 + yij**2)
         # ignore interactions for r > rcut
         if rij > rcut: continue
         # add contribution
         invr6 = 1. / rij**6
         invr12 = invr6 * invr6
         Epot += 4. * (invr12 - invr6)
   # return 1) total energy, and 2) temperature estimate T = Ekin / N
   return Ekin + Epot, Ekin / N


# plot particles, highlight 1st and 2nd particle
def plotParticles(ax0, r, L):
   ax0.set_xlim( [0, L] )
   ax0.set_ylim( [0, L] )
   #
   ax0.scatter(r[:,0], r[:,1], s = pointarea, color = "red")     # used red, except
   ax0.scatter(r[0,0], r[0,1], s = pointarea, color = "blue")    # particle 1 is in blue 
   ax0.scatter(r[1,0], r[1,1], s = pointarea, color = "green")   # particle 2 is in green






# read parameters
#
N    = int(   input('Number of particles: ') )
L    = float( input('Side length L of box [L/sigma units]: ') )
dt   = float( input('Time step [reduced units, see p.274]: ') )
plotFreq = int(   input('Number of timesteps between plots/output: ') )
vmax = float( input('Max initial speed [reduced units]: ') )
dmax = float( input('Max initial displacement rel. to grid sites: ') )
fname  = input('output file name: ').strip()

rcut = 3.   # cutoff for interaction: r > 3

params = (N, L, dt, rcut)


# open output file (truncate)
# 
if fname != "":   f = open(fname, "w+")
else:             f = None


# initial conditions
#

r     = np.zeros( (N, 2) )   # current position
rprev = np.zeros( (N, 2) )   # prev position
v     = np.zeros( (N, 2) )   # current velocity

# arrange particles on a rectangular grid, to keep them apart initially
# - use smallest grid that can house them
# - grid only partially filled if N is not square
Ngrid = int(sqrt(N))
if Ngrid != sqrt(N):
   Ngrid += 1
Lgrid = L / Ngrid

for k in range(N):
   # row, column and corresponding position (sqrt(2) factors maintain parity with old Matlab version)
   i, j = k // Ngrid,  k % Ngrid
   r[k,0] = (i + 0.5) * Lgrid + (random() - 0.5) * dmax * Lgrid * sqrt(2.)
   r[k,1] = (j + 0.5) * Lgrid + (random() - 0.5) * dmax * Lgrid * sqrt(2.)
   # velocity
   v[k,0] = vmax * (random() - 0.5) * sqrt(2.)
   v[k,1] = vmax * (random() - 0.5) * sqrt(2.)
   # previous position (assumes no interaction before t = 0)
   rprev[k,0] = r[k,0] - v[k,0] * dt
   rprev[k,1] = r[k,1] - v[k,1] * dt



# display initial configuration
#

fontsize = 30   # fontsize, need to be tuned to screen resolution
linewidth = 3
pointarea = 80. * (20. / N)**2   # scale to N = 20 

fig, axArray = plt.subplots(1, 1)
(ax0) = axArray

ax0.tick_params(labelsize = fontsize)
plotParticles(ax0, r, L)


# turn on interactive mode, so plt.show() does not block, then create plot window
matplotlib.pyplot.ion()
plt.show()


plt.pause(0.001)
input('[ENTER]')


# interactive loop
# - also records physics output

print("Type: 'c' to clear tracks")
print("      'd' to clear and disable green tracks")
print("      'u' to unable green tracks")
print("      'q' to move on to analysis plots")
print("      '+/-' to decreas velocities by +50%/-50%")
print("      '1/2/3/4' to increase velocities by 10, 20, 30, 40%")
print("      anything else for another batch of position updates")

output = []

t = 0.
doPlot = True

while True:

   # compute next plotFreq timesteps
   for _ in range(plotFreq):
      r, rprev, v = update(r, rprev, params)
   t += plotFreq * dt
   # physics output
   # - energy, temperature at now previous timestep
   E, T = calculateE(rprev, v, params)
   # - |r|^2 for particle 1, and |r1-r2|^2 for particles 1 & 2
   r1sq  = rprev[0,0]**2 + rprev[0,1]**2
   r12sq = (rprev[0,0] - rprev[1,0])**2 + (rprev[0,1] - rprev[1,1])**2
   output.append( [t, E, T, r1sq, r12sq] )
   if f != None:
      f.write(str(t) + " " + str(E) + " " + str(T) + " " + str(r1sq) + " " + str(r12sq) + "\n")
      f.flush()
   # plot positions
   if doPlot:
      ax0.scatter(r[:,0], r[:,1], s = pointarea, color = "green", facecolors = "none", linewidth = 1)
      titleStr = "t=" + str(round(t, 4)) + ", E=" + str(round(E,4)) + ", T=" + str(round(T,4))
      ax0.set_title(titleStr, fontsize = fontsize)
      plt.pause(0.001)

   # user input on what to do next
   inp = input("input [q,c,d,u,+,-,1,2,3,4]: ").strip()
   scale = None

   if inp == "q":
      break
   elif inp == "c" or inp == "d":
      ax0.clear()
      plotParticles(ax0, r, L)
      plt.pause(0.001)
      if inp == "d":   doPlot = False
   elif inp == "u":    doPlot = True
   elif inp == "+":    scale = 1.5
   elif inp == "-":    scale = 0.5
   elif inp == "1":    scale = 1.1
   elif inp == "2":    scale = 1.2
   elif inp == "3":    scale = 1.3
   elif inp == "4":    scale = 1.4

   # scale velocities upon request
   if scale != None:
      for k in range(N):
         rprev[k,0] = r[k,0] - scale * (r[k,0] - rprev[k,0])
         rprev[k,1] = r[k,1] - scale * (r[k,1] - rprev[k,1])



# close file
if f != None:  f.close()


# time series plots
#

output = np.array(output)

# E vs t
ax0.clear()
ax0.set_xlabel('t [red. units]', fontsize = fontsize)
ax0.set_ylabel('E [red. units]', fontsize = fontsize)
ax0.set_title('Energy vs time', fontsize = fontsize)
ax0.plot(output[:,0], output[:,1], linewidth = linewidth, color = "blue")

plt.pause(0.01)
input("[ENTER]")

# T vs t
ax0.clear()
ax0.set_xlabel('t [red. units]', fontsize = fontsize)
ax0.set_ylabel('T [red. units]', fontsize = fontsize)
ax0.set_title('Temperature (estimated) vs time', fontsize = fontsize)
ax0.plot(output[:,0], output[:,2], linewidth = linewidth, color = "blue")

plt.pause(0.01)
input("[ENTER]")

# |r1|^2 vs t
ax0.clear()
ax0.set_xlabel('t [red. units]', fontsize = fontsize)
ax0.set_ylabel('|r1|^2 [red. units]', fontsize = fontsize)
ax0.set_title('Position-squared vs time', fontsize = fontsize)
ax0.plot(output[:,0], output[:,3], linewidth = linewidth, color = "blue")

plt.pause(0.01)
input("[ENTER]")

# |r1-r2|^2 vs t
ax0.clear()
ax0.set_xlabel('t [red. units]', fontsize = fontsize)
ax0.set_ylabel('|r1-r2|^2 [red. units]', fontsize = fontsize)
ax0.set_title('Relative distance squared vs time', fontsize = fontsize)
ax0.plot(output[:,0], output[:,4], linewidth = linewidth, color = "blue")

plt.pause(0.01)
input("[ENTER]")

#EOF
